### DISCUSSION TITLE
Pretraining

### DISCUSSION THREAD

**Discussion by IHumonen** [mir-group/nequip#235] [Q&A] [answered]
Hello!
I have a model (allegro), trained via nequip, and now I want to retrain this model on another dataset. Is it possible to do such a thing using nequip? Thanks in advance for the answer.

**Comment by Linux-cpp-lisp** [comment, marked as answer]
Hi @IHumonen ,

This is definitely possible, and is one of the main uses of the `initialize_from_state` model builder (see docstring https://github.com/mir-group/nequip/blob/b62c4f4c41c1e1da6df560ffbe7d76eaed1d71e7/nequip/model/_weight_init.py#L13).

Basically:

1. Copy your config file and give it a new run name for the new training (the finetuning)
2. Change the dataset and any other trainer settings that you want to update for your finetuning/further training
3. Add the `initialize_from_state` model builder to the end of your `model_builders` list
4. Add the `initial_model_state` key to your config and set it to the path to a state dict of saved weights from which to initialize the model, such as the `best_model.pth` weights from your previous training run:
    ```
    initial_model_state: /path/to/pretraining/session/best_model.pth
    ```

**Reply by IHumonen** [reply]
Thank's a lot! It works

**Reply by Linux-cpp-lisp** [reply]
Great! Two other things I should mention:
 1. This is written in such a way that restarts / loading of the fine-tuned model will work 👍 
 2. Be careful not to change any _model_ hyperparameters in the finetuning YAML that would affect the meaning of the weights. Most such changes would lead to loud errors, but it's just something I want to note here for people who read this discussion.  (For example you could probably make the radial fully-connected networks have one less layer and only get a warning, but you'd have a pretty insane initial network.)

**Reply by nisargjoshi10** [reply]
@Linux-cpp-lisp for this:

```
Add the initialize_from_state model builder to the end of your model_builders list
```
by `model_builders` list are you referring to change the one in `nequip/model/_build.py` or  in config?

**Reply by Linux-cpp-lisp** [reply]
@nisargjoshi10 I mean in the config file. Generally, unless I explicitly say otherwise, I always mean to make a change in the config file (most likely) or in some custom components (model builder, module, loss function, etc.) that you are writing _outside_ the `nequip` code itself. 

The `nequip` code is designed to make most things we want to do possible in a modular fashion that doesn't require changes to the main package. Obviously they will occasionally be necessary, especially for more obscure tasks, but in general the goal is to avoid that kind of ad-hoc "100 different versions of the code for 100 different projects" case.

**Reply by nisargjoshi10** [reply]
okay thanks! just wanted to make sure as I didn't see the slurm output mentioning about the `initialize_from_state`or weights from the previous run

**Reply by Linux-cpp-lisp** [reply]
@nisargjoshi10 that's a good point, I've added a TODO to add some logging to those so it isn't silent.

**Reply by terryfrankcombe** [reply]
Hi @Linux-cpp-lisp.  I don't see `model_builders` (or `initial_model_state`) in configs/full.yaml.  Is that an oversight?

**Reply by Linux-cpp-lisp** [reply]
Hi @terryfrankcombe ,

Ah yeah we usually don't specify that for NequIP models, but it is required for Allegro, so the original poster already had it in his config. I could add it; do you think it would be clearer?

**Reply by terryfrankcombe** [reply]
I think it should be documented _somewhere_.  I found an example in a different thread (and so seems useful to a non-empty subset of people), but it wasn't obvious what should be there otherwise.

**Reply by Linux-cpp-lisp** [reply]
@terryfrankcombe I've added the default `model_builders` with some explaination to `full.yaml`: https://github.com/mir-group/nequip/commit/4b2ef8132e46cb9d5c9828160b3477f821da6149

Thanks for the feedback on this.

**Reply by wwwjjj123yyyy** [reply]
Hi, I want to ask a qestion that when I followed the steps you showed to begin a new work with different batch size and lr_scheduler_patience in the example.yaml file using the previous model: best_model.pth. The work ends with the warning"AttributeError: 'NoneType' object has no attribute 'irreps_out'". I don't know why.

**Reply by Linux-cpp-lisp** [reply]
@wwwjjj123yyyy can you please post a complete stack trace?

**Reply by wwwjjj123yyyy** [reply]
This is the bottom lines of my output file when running nequip,  I hope it helps.
![Uploading 微信截图_20231121132515.png…]()

**Reply by Linux-cpp-lisp** [reply]
I can't open the file; can you please provide it as text here in GitHub?

**Reply by wwwjjj123yyyy** [reply]
@Linux-cpp-lisp  This is my otuput file.
[slurm-286959.txt](https://github.com/mir-group/nequip/files/13433899/slurm-286959.txt)

**Reply by Mathieu-Istas** [reply]
Hi, I do have the same error as @wwwjjj123yyyy . I copied a config I used to train a model and wanted to do some training with some extra data with it, so I added the lines
```
### start from a previous run
model_builders:
 - initialize_from_state
initial_model_state: path_to_previous_model/best_model.pth
```
And after processing the data I got the error
```
Traceback (most recent call last):
  File "/home/istas/.local/bin/nequip-train", line 8, in <module>
    sys.exit(main())
  File "/home/istas/.local/lib/python3.9/site-packages/nequip/scripts/train.py", line 83, in main
    trainer = fresh_start(config)
  File "/home/istas/.local/lib/python3.9/site-packages/nequip/scripts/train.py", line 211, in fresh_start
    final_model = model_from_config(
  File "/home/istas/.local/lib/python3.9/site-packages/nequip/model/_build.py", line 141, in model_from_config
    model = GraphModel(
  File "/home/istas/.local/lib/python3.9/site-packages/nequip/nn/_graph_model.py", line 47, in __init__
    self._init_irreps(irreps_in=irreps_in, irreps_out=model.irreps_out)
AttributeError: 'NoneType' object has no attribute 'irreps_out'
```
Should I specify `irreps_out` manually in the config file? Something like `chemical_embedding_irreps_out: 32x0e`?

**Comment by IZugec** [comment]
@Linux-cpp-lisp 
Firstly, thanks for clarifications and the discussion. I wanted to talk about the statement you have written above:

"Be careful not to change any model hyperparameters in the finetuning YAML that would affect the meaning of the weights."

I wanted to ask you specifically for the radius which is obviously extremely important (here I am retraining nequip). Do you think it's safe to change radius as a hyperparameter?

**Reply by Linux-cpp-lisp** [reply]
Hi @IZugec ,

Let me make sure I understand what you're asking before I answer: you want to train a model with some `r_max`, and then fine-tune it later with a different `r_max`?  Can you explain a little bit the motivation, if this is the case, because it doesn't make a ton of sense to me on its own.

**Reply by IZugec** [reply]
You understood correctly. I want to fine-tune it on different dataset which I think would benefit from larger r_max, but want to continue from weights that was learned on this initial dataset.

**Reply by Linux-cpp-lisp** [reply]
I see... this is complicated by the radial basis projection: our networks don't process distances directly, but rather project them out into a radial basis (we use the Bessel basis, see the papers for details and references).  This radial basis depends on `r_max` and the number of radial basis functions, so if you change `r_max` the basis changes and thus completely changes the interpretation of the weights.

The first approach that comes to my mind would be to set a larger `r_max` for the radial basis in the original model using:
```
BesselBasis_r_max: 10.0  # or whatever
```
but still setting a smaller overall `r_max`. This way the original dataset and smooth cutoff will be set to your smaller cutoff.  Then, when you want to fine tune, you also set `r_max` to 10.0 (or whatever), but the radial basis won't change--- only the polynomial cutoff and the dataset preprocessing cutoff, so the fine-tuning _should_ go OK...

**Reply by IZugec** [reply]
Hmm I see. Thank you very much for your time to reply. Appreciate the insight.

**Comment by ken1944** [comment]
Any suggestions to solve `CUDA out of memory` error while finetuning? I am using a batch size of 1.

**Reply by Linux-cpp-lisp** [reply]
Is there a specific reason you think this is related to finetuning vs training in general? Can you please open a new discussion thread with a full description and traceback?

**Reply by ken1944** [reply]
Sure, made a new one here #388
